### PR TITLE
include resource additional metadata as web app keys

### PR DESCRIPTION
**Pull request by sblack-usu**
### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [x] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. Add a custom key to a web app url.  Mark all sharing permissions and Composite Resource type.  Add the web app to your open with list.  Go to a composite resource, validate the web app does not show.  Add additional metadata with the same key as the custom key provided in the web app and anything you'd like as the value.  Refresh the resource page and validate the web app appears.  Click the web app from the list and validate the url replaces the custom key with the value provided in your reosurce.

### CODE REVIEW

**Comment by sblack-usu** [comment]
@Castronova - would you be interested in doing the user documentation for this feature.  I think you could write better documentation on this from a user's perspective than I could.  https://help.hydroshare.org/apps/web-app/

File: hs_tools_resource/tests/test_web_app.py [hunk 1]
```diff
@@ -371,6 +371,46 @@ def test_web_app_extended_metadata_appkey_association(self):
         self.assertIsNone(relevant_tools, msg='relevant_tools should be None with no appkey '
                                               'matching')
 
+    def test_web_app_extended_metadata_custom_key(self):
+        # testing a resource can be associated with a web app tool resource via
+        # appkey name-value extended metadata matching
```
Comment (hyi) [inline hs_tools_resource/tests/test_web_app.py:376]: This comment does not appear to apply to this test method.

File: hs_core/models.py [hunk 2]
```diff
@@ -3498,6 +3498,7 @@ def get_hs_term_dict(self):
 
         hs_term_dict["HS_RES_ID"] = self.short_id
         hs_term_dict["HS_RES_TYPE"] = self.resource_type
+        hs_term_dict.update(self.extra_metadata.items())
```
Comment (sblack-usu) [inline hs_core/models.py:3501]: @hyi - here is the functional change.   It's small and easy to miss.

Comment (hyi) [reply hs_core/models.py:3501]: @sblack-usu Got it. I actually did not miss this line, but just did not relate this to the functional change. Now it makes sense to me. I'll approve the PR.

**Comment by sblack-usu** [comment]
@hyi - I just added another feature to this PR allowing a web app to specify a default value for a custom web app key in the web app's additional.  This was added along with tests looking for the default value as well as overriding the default value by specifying the key in the resource's additional metadata.

**Comment by Castronova** [comment]
> @Castronova - would you be interested in doing the user documentation for this feature. I think you could write better documentation on this from a user's perspective than I could. https://help.hydroshare.org/apps/web-app/

@sblack-usu I can take a crack at it.

**Comment by sblack-usu** [comment]
let me know if you have any questions or if you want a quick demo let me know.

**Comment by sblack-usu** [comment]
Any updates on this @Castronova?

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|
| NEW CODE SMELLS | 3 |
| NEW VULNERABILITIES | 0 |
| NEW SECURITY HOTSPOTS | 0 |
| NEW DUPLICATED LINES DENSITY | 5.08474576271186 |
| NEW BUGS | 5 |

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-3850](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-3850)

**Pull request merged**